### PR TITLE
Implement event pod handler

### PR DESCRIPTION
**Pull request by Mmduh-483**

### CODE REVIEW

File: pkg/watcher/resouce-event-handler/resource_event_handler.go [hunk 1]
```diff
@@ -1,11 +1,14 @@
...
+	"github.com/Mellanox/ib-kubernetes/pkg/utils"
+
 	"k8s.io/apimachinery/pkg/runtime"
 	"k8s.io/client-go/tools/cache"
 )
 
 type ResourceEventHandler interface {
 	cache.ResourceEventHandler
 	GetResourceObject() runtime.Object
+	GetResult() (*utils.SynchronizedMap, *utils.SynchronizedMap)
```
Comment (moshe010) [inline pkg/watcher/resouce-event-handler/resource_event_handler.go]: GetResults

File: pkg/daemon/daemon.go [hunk 2]
```diff
@@ -21,7 +21,11 @@ type daemon struct {
 // It returns error in case of failure.
 func NewDaemon() (Daemon, error) {
 	glog.Info("daemon NewDaemon():")
-	podEventHandler := resEvenHandler.NewPodEventHandler()
+	kubeClient, err := k8sClient.NewK8sClient()
+	if err != nil {
+		return nil, err
+	}
+	podEventHandler := resEvenHandler.NewPodEventHandler(kubeClient)
 	client, err := k8sClient.NewK8sClient()
```
[the remark this answers is not in the document]
Comment (Mmduh-483) [reply pkg/daemon/daemon.go]: removed

File: pkg/utils/utils.go [hunk 3]
```diff
@@ -13,3 +19,35 @@ func PodWantsNetwork(pod *kapi.Pod) bool {
...
+}
+
+// PodWantsInfiniBand check if pod is needs infiniband support
+func PodWantsInfiniBand(pod *kapi.Pod) bool {
+	_, ok := pod.Annotations[InfiniBandAnnotation]
+	return ok
+}
+
+// PodIsConfiguredWithInfiniBand check if pod is already infiniband supported
+func PodIsConfiguredWithInfiniBand(pod *kapi.Pod) bool {
```
[the remark this answers is not in the document]
Comment (Mmduh-483) [reply pkg/utils/utils.go]: changed

File: pkg/watcher/resouce-event-handler/pod.go [hunk 4]
```diff
@@ -1,32 +1,242 @@
...
+	addedPods   *utils.SynchronizedMap
+	deletedPods *utils.SynchronizedMap
 }
 
-func NewPodEventHandler() ResourceEventHandler {
-	return &podEventHandler{}
+func NewPodEventHandler(kubeClient k8sClient.Client) ResourceEventHandler {
+	eventHandler := &podEventHandler{
+		retryPods:   sync.Map{},
+		kubeClient:  kubeClient,
```
[the remark this answers is not in the document]
Comment (Mmduh-483) [reply pkg/watcher/resouce-event-handler/pod.go]: removed

File: pkg/watcher/resouce-event-handler/pod.go [hunk 5]
```diff
@@ -1,32 +1,242 @@
...
+	}
+
+	glog.V(3).Infof("OnDelete(): successfully deleted namespace %s name %s", pod.Namespace, pod.Name)
+}
+
+func (p *podEventHandler) GetResult() (*utils.SynchronizedMap, *utils.SynchronizedMap) {
+	return p.addedPods, p.deletedPods
+}
+
+func (p *podEventHandler) setConfiguredPod(pod *kapi.Pod) error {
```
[the remark this answers is not in the document]
Comment (Mmduh-483) [reply pkg/watcher/resouce-event-handler/pod.go]: removed, will be added in the periodic update PR

File: pkg/watcher/resouce-event-handler/pod.go [hunk 6]
```diff
@@ -1,32 +1,242 @@
...
+		// skip repeat request for same network
+		if podNetworks[network.Name] {
+			glog.Warningf("OnAdd(): Repeated request for network is not allowed, pod %s, network %s",
+				pod.Name, network.Name)
+			continue
+		}
+		podNetworks[network.Name] = true
+
+		//check if pod network has guid
+		if !utils.NetworkHasGuid(network) {
```
[the remark this answers is not in the document]
Comment (Mmduh-483) [reply pkg/watcher/resouce-event-handler/pod.go]: by mistake, it skipped if it is configured

File: pkg/watcher/resouce-event-handler/pod.go [hunk 7]
```diff
@@ -1,32 +1,242 @@
...
+		if !ok {
+			pods = []*kapi.Pod{pod}
+		} else {
+			pods = append(pods.([]*kapi.Pod), pod)
+		}
+
+		p.addedPods.Set(network.Name, pods)
+	}
+
+	if err := p.setConfiguredPod(nil); err != nil {
```
[the remark this answers is not in the document]
Comment (Mmduh-483) [reply pkg/watcher/resouce-event-handler/pod.go]: removed

File: pkg/watcher/resouce-event-handler/pod.go [hunk 8]
```diff
@@ -1,32 +1,242 @@
...
-	glog.Infof("OnUpdate(): oldPod %v, newPod %v", oldObj, newObj)
+	glog.V(4).Infof("OnUpdate(): oldPod %v, newPod %v", oldObj, newObj)
+	pod := newObj.(*kapi.Pod)
+	glog.Infof("OnUpdate(): namespace %s name %s", pod.Namespace, pod.Name)
+
+	if !utils.PodWantsNetwork(pod) || !utils.PodScheduled(pod) {
+		return
+	}
+
+	if utils.PodIsConfiguredWithInfiniBand(pod) {
```
[the remark this answers is not in the document]
Comment (Mmduh-483) [reply pkg/watcher/resouce-event-handler/pod.go]: check

File: pkg/watcher/resouce-event-handler/resource_event_handler.go (hunk 1 above)
Comment (Mmduh-483) [reply pkg/watcher/resouce-event-handler/resource_event_handler.go]: renamed

File: pkg/utils/utils.go [hunk 9]
```diff
@@ -13,3 +19,35 @@ func PodWantsNetwork(pod *kapi.Pod) bool {
...
+	return len(pod.Annotations[v1.NetworkAttachmentAnnot]) > 0
+}
+
+// PodIsRunning check if pod is in "Running" state
+func PodIsRunning(pod *kapi.Pod) bool {
+	return pod.Status.Phase == kapi.PodRunning
+}
+
+// PodWantsInfiniBand check if pod is needs infiniband support
+func PodWantsInfiniBand(pod *kapi.Pod) bool {
```
[the remark this answers is not in the document]
Comment (Mmduh-483) [reply pkg/utils/utils.go]: removed

File: pkg/utils/utils.go [hunk 10]
```diff
@@ -13,3 +21,44 @@ func PodWantsNetwork(pod *kapi.Pod) bool {
...
+	return len(pod.Annotations[v1.NetworkAttachmentAnnot]) > 0
+}
+
+// PodIsRunning check if pod is in "Running" state
+func PodIsRunning(pod *kapi.Pod) bool {
+	return pod.Status.Phase == kapi.PodRunning
+}
+
+// ParseInfiniBandAnnotation return a map of InfiniBand annotation
+func ParseInfiniBandAnnotation(pod *kapi.Pod) map[string]string {
```
Comment (moshe010) [inline pkg/utils/utils.go]: I think it is better to return (map[string[string , error)

File: pkg/utils/utils.go [hunk 11]
```diff
@@ -13,3 +21,44 @@ func PodWantsNetwork(pod *kapi.Pod) bool {
...
+	if err := json.Unmarshal([]byte(value), &ibAnnotation); err != nil {
+		return nil
+	}
+
+	return ibAnnotation
+}
+
+// PodIsConfiguredWithInfiniBand check if pod is already infiniband supported
+func IsPodNetworkConfiguredWithInfiniBand(ibAnnotation map[string]string, network string) bool {
+	return ibAnnotation != nil && ibAnnotation[network] == ConfiguredInfiniBandPod
```
Comment (moshe010) [inline pkg/utils/utils.go]: what if network is not exist in Annotation?

File: pkg/watcher/resouce-event-handler/pod.go [hunk 12]
```diff
@@ -1,32 +1,190 @@
...
+		return
+	}
+
+	for _, network := range networks {
+		if !utils.IsPodNetworkConfiguredWithInfiniBand(ibAnnotation, network.Name) {
+			continue
+		}
+
+		//check if pod network has guid
+		if !utils.NetworkHasGuid(network) {
```
Comment (moshe010) [inline pkg/watcher/resouce-event-handler/pod.go]: s/NetworkHasGuid/PodNetworkHasGuid

File: pkg/watcher/resouce-event-handler/pod.go [hunk 13]
```diff
@@ -1,32 +1,190 @@
...
+	ibAnnotation := utils.ParseInfiniBandAnnotation(pod)
+
+	for _, network := range networks {
+		// check if pod network is configured
+		if utils.IsPodNetworkConfiguredWithInfiniBand(ibAnnotation, network.Name) {
+			continue
+		}
+
+		//check if pod network has guid
+		if utils.NetworkHasGuid(network) {
```
Comment (moshe010) [inline pkg/watcher/resouce-event-handler/pod.go]: remove this if he has guid we still need to configure it. guid can be passed from user

File: pkg/utils/utils.go (hunk 10 above)
Comment (Mmduh-483) [reply pkg/utils/utils.go]: changed

File: pkg/utils/utils.go (hunk 11 above)
Comment (Mmduh-483) [reply pkg/utils/utils.go]: then it return false

File: pkg/watcher/resouce-event-handler/pod.go (hunk 12 above)
Comment (Mmduh-483) [reply pkg/watcher/resouce-event-handler/pod.go]: renamed

File: pkg/watcher/resouce-event-handler/pod.go (hunk 13 above)
Comment (Mmduh-483) [reply pkg/watcher/resouce-event-handler/pod.go]: removed

File: pkg/utils/utils.go (hunk 11 above)
Comment (Mmduh-483) [reply pkg/utils/utils.go]: added

**Pull request merged**